internal/docker.go

```diff
@@ -0,0 +1,296 @@
+}
+
+// DockerClient offers some methods for querying docker. You must use the
+// NewDockerClient() method to make one, or the methods won't work.
+type DockerClient struct {
+	client             DockerInteractor
+	existingContainers map[string]bool
+}
+
+// NewDockerClient creates a new DockerClient, connecting to docker using
+// the standard environment variables to define options.
+// Refer: https://github.com/moby/moby/blob/1.13.x/client/client.go
+// eg of creating in production code...
+/*
+	import	"github.com/docker/docker/client"
+	....
+	cli, err := client.NewEnvClient()
+	if err != nil {
+		panic(err)
+	}
+
+	dockerClient := &DockerClient{
+		client:             cli,
+		existingContainers: make(map[string]bool),
+	}
+
+	cntrList, err := dockerClient.GetCurrentContainers(ctx)
+	if err != nil {
+		panic(err)
+	}
+	fmt.Println(cntrList)
+*/
+func NewDockerClient(docInteractor DockerInteractor) *DockerClient {
+	return &DockerClient{
+		client:             docInteractor,
+		existingContainers: make(map[string]bool),
+	}
+}
+
+// GetCurrentContainers returns current containers.
+func (d *DockerClient) GetCurrentContainers(ctx context.Context) ([]types.Container, error) {
+	return d.client.ContainerList(ctx, types.ContainerListOptions{})
+}
+
+// RememberCurrentContainerIDs calls GetCurrentContainerIDs() and stores the
+// results, for the benefit of a future GetNewDockerContainerID() call.
+func (d *DockerClient) RememberCurrentContainerIDs(ctx context.Context) error {
+	containers, err := d.GetCurrentContainers(ctx)
+	if err != nil {
+		return &ContainerError{Type: ErrContainerList, Err: err}
+	}
+
+	for _, container := range containers {
+		d.existingContainers[container.ID] = true
+	}
+
+	return nil
+}
+
+// GetNewDockerContainerIDs returns the ids of all the docker containers that now exist
+// and weren't previously remembered by a RememberCurrentContainerIDs() call.
+func (d *DockerClient) GetNewDockerContainerIDs(ctx context.Context) ([]string, error) {
+	containers, err := d.GetCurrentContainers(ctx)
+	if err != nil {
+		return []string{}, &ContainerError{Type: ErrContainerList, Err: err}
+	}
+
+	newCntrsIDs := []string{}
+
+	for _, container := range containers {
+		if !d.existingContainers[container.ID] {
+			newCntrsIDs = append(newCntrsIDs, container.ID)
+		}
+	}
+
+	if len(newCntrsIDs) == 0 {
+		return []string{}, &ContainerError{Type: ErrNoNewContainerFound}
+	}
+
+	return newCntrsIDs, nil
+}
+
+// GetNewDockerContainerIDByName returns the id of the container with given name.
+// Only new containers (those not remembered by the last RememberCurrentContainerIDs() call)
+// are considered when searching for a container with that name.
+func (d *DockerClient) GetNewDockerContainerIDByName(ctx context.Context, name string) (string, error) {
+	containers, err := d.GetCurrentContainers(ctx)
+	if err != nil {
+		return "", &ContainerError{Type: ErrContainerList, Err: err}
+	}
+
+	for _, container := range containers {
```

Comment:
This should loop through the output of GetNewDockerContainerIDs(), so the code above isn't needed, and the next line isn't needed.

Reply:
Created a new function GetNewDockerContainers() to solve this issue which is also useful in case we want to get all the new containers (not just the ids)